src/module_3/module3_statistical_learning.md

```diff
@@ -0,0 +1,1368 @@
+```
+
+
+```python
+def split_by_order_id(df, test_size=0.1, validation_size=0.1, random_state=42):
+    """
+    Splits a DataFrame into train, validation, and test sets based on unique order IDs.
+
+    Args:
+        df (pd.DataFrame): The input DataFrame.
+        test_size (float, optional): Proportion of the dataset to include in the test set.
+                                     Defaults to 0.2.
+        validation_size (float, optional): Proportion of the dataset to include in the validation set.
+                                           Defaults to 0.2.
+        random_state (int, optional): Controls the randomness of the split. 
+                                      Set for reproducible results. Defaults to None.
+
+    Returns:
+        tuple: A tuple containing three DataFrames: (train_df, validation_df, test_df)
+    """
+
+    # Get unique order IDs
+    unique_order_ids = df['order_id'].unique()
+
+    # Split order IDs into train and test groups 
+    train_order_ids, test_order_ids = train_test_split(unique_order_ids, 
+                                                       test_size=test_size, 
+                                                       random_state=random_state)
+
+    # Optionally, split train order IDs further into train and validation groups
+    if validation_size > 0:
+        train_order_ids, validation_order_ids = train_test_split(train_order_ids,
+                                                                 test_size=validation_size / (1 - test_size),
+                                                                 random_state=random_state)
+
+    # Create split masks
+    train_mask = df['order_id'].isin(train_order_ids)
+    validation_mask = df['order_id'].isin(validation_order_ids)
+    test_mask = df['order_id'].isin(test_order_ids)
+
+    # Split the DataFrame based on masks
+    train_df = df[train_mask]
+    validation_df = df[validation_mask]
+    test_df = df[test_mask]
+
+    return train_df, validation_df, test_df
+
+train_df_0, validation_df_0, test_df_0 = split_by_order_id(df_orders)
+```
+
+### Create a baseline model
+
+
+```python
+train_df = train_df_0.copy()
+validation_df = validation_df_0.copy()
+
+### Label Encoder
+encoders = {}
+for col in categorical_cols:
+    le = LabelEncoder()
+    train_df[col] = le.fit_transform(train_df[col])
+    encoders[col] = le
+
+# Save the encoding to reuse it in validation and test
+with open(data_path + 'label_encoders.pkl', 'wb') as f:
+    pickle.dump(encoders, f)
+
+# ------------------ Model Training ------------------
+    
+X_train = train_df.drop(label_col, axis=1)
+y_train = train_df[label_col]
+
+# Fit the LogisticRegression model
+log_reg = LogisticRegression() 
+log_reg.fit(X_train, y_train) 
+
+# ------------------ Classification ------------------
+
+# Encode test data
+with open(data_path + 'label_encoders.pkl', 'rb') as f:
+    encoders = pickle.load(f)
+
+for col in categorical_cols: 
```

Comment:
put this into a function or even better a pipeline: https://scikit-learn.org/stable/modules/generated/sklearn.pipeline.Pipeline.html